File: JS OOP/Modules and Patterns/student.js
'use strict';

class Student {
    constructor(firstName, lastName) {
        this.firstName = firstName;
        this.lastName = lastName;
        this._id = this._firstName.length % 9;
        this._submittedHomeworks = 0;
        this._score = 0;
    }

    get firstName() {
        return this._firstName;
    }

    set firstName(firstName) {
        if (!/^[A-Z][a-z]*$/g.test(firstName)) {
            throw 'Invalid student first name';
        }

        this._firstName = firstName;
    }

    get lastName() {
        return this._lastName;
    }

    set lastName(lastName) {
        if (!/^[A-Z][a-z]*$/g.test(lastName)) {
            throw 'Invalid studnet last name';
        }

        this._lastName = lastName;
    }

    get id() {
        return this._id;
    }

    get submittedHomeworks() {
        return this._submittedHomeworks;
    }

    get score() {
        return this._score;
    }

    addScore(score) {
        this._score += score;
    }

    submitHomework() {
        this._submittedHomeworks++;
    }
}

module.exports = Student;

File: High Quality Code/HighQualityMethods/TaskOne/Student.cs
namespace TaskOne
{
    using System;

    public class Student
    {
        private const string DefaultHobby = "None";

        private string firstName;
        private string lastName;
        private string otherInfo;

        public Student(string firstName, string lastName)
        {
            this.FirstName = firstName;
            this.LastName = lastName;
        }

        public string FirstName
        {
            get
            {
                return this.firstName;
            }

            set
            {
                if (string.IsNullOrEmpty(value))
                {
                    throw new ArgumentException("Invalid first name.");
                }

                this.firstName = value;
            }
        }

        public string LastName
        {
            get
            {
                return this.lastName;
            }

            set
            {
                if (string.IsNullOrEmpty(value))
                {
                    throw new ArgumentException("Invalid last name.");
                }

                this.lastName = value;
            }
        }

        public string OtherInfo
        {
            get
            {
                return this.otherInfo;
            }

            set
            {
                if (string.IsNullOrEmpty(value))
                {
                    throw new ArgumentException("Student other info cannot be null or an empty string.");
                }

                this.otherInfo = value;
            }
        }

        public bool IsOlderThan(Student otherStudent)
        {
            DateTime firstDate = DateTime.Parse(this.OtherInfo.Substring(this.OtherInfo.Length - 10));
            DateTime secondDate = DateTime.Parse(otherStudent.OtherInfo.Substring(otherStudent.OtherInfo.Length - 10));

            return firstDate > secondDate;
        }
    }
}
